File: src/test/java/com/thebluehats/server/game/enchants/HealerTest.java
package com.thebluehats.server.game.enchants;

import com.thebluehats.server.game.managers.combat.templates.EventTemplate;
import com.thebluehats.server.game.managers.combat.templates.PlayerHitPlayerTemplate;
import org.bukkit.attribute.Attribute;
import org.bukkit.attribute.AttributeInstance;
import org.bukkit.entity.Player;
import org.junit.Test;

import static org.mockito.Mockito.*;

public class HealerTest {
    @Test
    public void DamagerAndDamageeGetHealedWhenHit() {
        Healer enchant = new Healer(new EventTemplate[] { new PlayerHitPlayerTemplate() });
        Player damaged = mock(Player.class);
        Player damager = mock(Player.class);

        double maxHealth = 20;
        int healAmount = 4;

        when(damaged.getAttribute(Attribute.GENERIC_MAX_HEALTH)).thenReturn(mock(AttributeInstance.class));
        when(damager.getAttribute(Attribute.GENERIC_MAX_HEALTH)).thenReturn(mock(AttributeInstance.class));

        when(damaged.getAttribute(Attribute.GENERIC_MAX_HEALTH).getValue()).thenReturn(maxHealth);
        when(damager.getAttribute(Attribute.GENERIC_MAX_HEALTH).getValue()).thenReturn(maxHealth);

        enchant.execute(new HealerArgs(damager, damaged, healAmount));

        verify(damaged).setHealth(damaged.getHealth() + healAmount);
        verify(damager).setHealth(damager.getHealth() + healAmount);

        enchant.execute(new HealerArgs(damager, damaged, 22));

        verify(damaged).setHealth(maxHealth);
        verify(damager).setHealth(maxHealth);
    }
}

File: src/test/kotlin/com/thebluehats/server/game/enchants/HealerTest.kt
package com.thebluehats.server.game.enchants

import com.thebluehats.server.game.managers.enchants.processedevents.DamageEventEnchantData
import io.kotest.core.spec.style.DescribeSpec
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.bukkit.entity.Player

class HealerTest : DescribeSpec({
    describe("healer") {
        it("heals the damager and damagee on hit") {
            val healer = Healer(mockk())

            val damagerMock = mockk<Player> {
                every { maxHealth } returns 20.0
                every { health } returns 5.0
                every { health = any() } returns Unit
            }

            val damageeMock = mockk<Player> {
                every { maxHealth } returns 20.0
                every { health } returns 5.0
                every { health = any() } returns Unit
            }

            val data = mockk<DamageEventEnchantData> {
                every { level } returns 3
                every { damager } returns damagerMock
                every { damagee } returns damageeMock
            }

            healer.execute(data)

            verify { damagerMock.health = 11.0 }
            verify { damageeMock.health = 11.0 }
        }
    }
})
